File: docs/papers211_215_sequence/qa/p211_a_pair_binding/prepare_binding.py
"""Root A strict-pair approval after exact initial adoption; no science run."""
from hashlib import sha256
import json
import os
from pathlib import Path

ROOT = Path('/root/autodl-tmp/symbolic_dynamics')
QA = ROOT/'docs/papers211_215_sequence/qa'
OUT = Path(__file__).resolve().parent
INITIAL = QA/'p211_a_initial_binding'
ATTEMPT = QA/'root_replays/p211_a_pair_01'
READS = {}


def raw(path):
    path = Path(path)
    data = path.read_bytes()
    if str(path) in READS:
        assert READS[str(path)] == data
    READS[str(path)] = data
    return data


def identity(data):
    return {'sha256': sha256(data).hexdigest(), 'bytes': len(data)}


def pin(path, expected=None):
    path = Path(path)
    result = {**identity(raw(path)), 'resolved': str(path.resolve(strict=True)),
              'symlink': os.readlink(path) if path.is_symlink() else None}
    if expected is not None:
        assert all(result[key] == expected[key] for key in result if key in expected), str(path)
    return result


def doc(path):
    return json.loads(raw(path))


def write(name, value):
    with (OUT/name).open('xb') as stream:
        stream.write((json.dumps(value, sort_keys=True, indent=2, allow_nan=False)+'\n').encode())


assert OUT == QA/'p211_a_pair_binding' and not (OUT/'BINDING.json').exists()
assert not os.path.lexists(ATTEMPT) and ATTEMPT.parent.resolve(strict=True) == ATTEMPT.parent
old_bytes = raw(INITIAL/'BINDING.json')
assert sha256(old_bytes).hexdigest() == '8f019e9cc35ba0c0d3ed02d9b419f844acc09f0df83ae8aeff590811463b5f0f'
old = json.loads(old_bytes)
binding = json.loads(old_bytes)
assert old['approved'] is True and old['role'] == 'A' and old['mode'] == 'initial'
for row in doc(INITIAL/'INPUTS_AT_BINDING.json').items():
    pin(*row)
for name, status in [('PRODUCTION_RECORDS_NATIVE01.json', 'PASS_ROOT_COMPLETE_A_PRODUCTION_RECORDS'),
                     ('OUTPUT_SEMANTICS_NATIVE01.json', 'PASS_FULL_SAVED_A_OUTPUT_SEMANTICS'),
                     ('ADOPTION_NATIVE01.json', 'ROOT_A_CANONICAL_ADOPTED_FROM_COMPLETE_ACTUAL_INITIAL_STDOUT')]:
    native = doc(INITIAL/name)
    assert native['exit_code'] == 0 and json.loads(native['output'])['status'] == status
adoption = doc(INITIAL/'ADOPTION/RESULT.json')
canonical = Path(old['canonical']['path'])
canonical_pin = pin(canonical, adoption)
assert canonical_pin['sha256'] == 'e637aa186b5d11c3bba1ad5318d42c7dc93ea736fee9ed3013ebcc1beaa9cd43'
assert canonical_pin['bytes'] == 1313394 and raw(canonical) == raw(adoption['source'])
adoption_files = []
for line in raw(INITIAL/'ADOPTION/SHA256SUMS').decode().splitlines():
    digest, name = line.split('  ', 1)
    assert name != 'SHA256SUMS' and not Path(name).is_absolute() and '..' not in Path(name).parts
    pin(INITIAL/'ADOPTION'/name, {'sha256': digest})
    adoption_files.append(name)
assert len(adoption_files) == 6 and sorted(adoption_files+['SHA256SUMS']) == sorted(path.name for path in (INITIAL/'ADOPTION').iterdir())
comparison = doc(INITIAL/'ADOPTION/CMP_RECEIPT.json')
assert comparison['argv'] == ['/usr/bin/cmp', '--', adoption['source'], str(canonical)]
assert comparison['exit_code'] == 0 and comparison['cwd'] == str(ROOT)
assert comparison['environment'] == {'PATH': '/usr/bin:/bin', 'LANG': 'C.UTF-8', 'LC_ALL': 'C.UTF-8', 'TZ': 'UTC'}
assert doc(INITIAL/'ADOPTION/CMP_ATTEMPT.json') == {key: comparison[key] for key in doc(INITIAL/'ADOPTION/CMP_ATTEMPT.json')}
for name in ('stdout', 'stderr'):
    body = raw(INITIAL/'ADOPTION'/('cmp.'+name+'.raw'))
    assert body == b'' and identity(body) == comparison[name]
transcript = doc(canonical)
binding['mode'], binding['attempt'] = 'pair', str(ATTEMPT)
binding['canonical'] = {'path': str(canonical), **canonical_pin}
binding['schema']['equalities'].append({'path': ['check_total'], 'value': 30951})
for index, carrier in enumerate(transcript['carriers']):
    binding['schema']['equalities'].extend([
        {'path': ['carriers', index, 'check_counts'], 'value': carrier['check_counts']},
        {'path': ['carriers', index, 'check_total'], 'value': carrier['check_total']}])
additions = [INITIAL/name for name in ('BINDING.json', 'INPUTS_AT_BINDING.json',
            'PRODUCTION_RECORDS_NATIVE01.json', 'OUTPUT_SEMANTICS_NATIVE01.json',
            'PRODUCTION_TOOL_INVOCATION.json', 'ADOPTION_NATIVE01.json', 'adopt_initial_output.py')]
additions += list((INITIAL/'ADOPTION').iterdir())
additions += [Path(__file__).resolve(), QA/'p211_author_pair_binding/prepare_binding.py']
for path in sorted(additions):
    binding['provenance_inputs'].append({'path': str(path), **pin(path)})
binding['root_canonical_policy'] = 'Existing accepted A canonical is immutable; compare each full stdout to it and the two stdout files to each other; never publish or replace.'
assert all(type(value) is int and 0 < value <= 900 for value in binding['timeouts'].values())
changed = {key: {'before': old[key], 'after': binding[key]} for key in old if old[key] != binding[key]}
assert set(changed) == {'mode', 'attempt', 'canonical', 'schema', 'provenance_inputs', 'root_canonical_policy'}
for path, data in READS.items():
    assert Path(path).read_bytes() == data
write('INPUTS_AT_BINDING.json', {path: pin(path) for path in list(READS)})
write('BINDING.json', binding)
write('EXACT_INITIAL_TO_PAIR_DELTA.json', changed)
result = {'status': 'ROOT_A_STRICT_PAIR_BINDING_APPROVED_NOT_EXECUTED',
    'binding': {'path': str(OUT/'BINDING.json'), **pin(OUT/'BINDING.json')},
    'attempt': str(ATTEMPT), 'canonical': {'path': str(canonical), **canonical_pin},
    'changed_top_keys': sorted(changed), 'input_keys_checked': len(READS),
    'new_scientific_invocations': 0, 'required_scientific_invocations': 2,
    'required_native_canonical_pair_comparisons': 3,
    'infrastructure_reuse': 'accepted author strict-pair binding mechanics adapted to exact A schema and pins; all initial source/environment/schema contracts retained'}
write('RESULT.json', result)
print(json.dumps(result, sort_keys=True))

File: symbolic_dynamics/docs/papers211_215_sequence/qa/p212_author_pair_binding01/prepare_binding.js
'use strict';
// Root forward derivative of the fully read accepted initial binding preparer.
// No Python/AST/import/probe/science; old absent-canonical checker is not rerun.
const fs=require('fs'),path=require('path'),crypto=require('crypto'),a=require('assert/strict');
const ROOT='/root/autodl-tmp/symbolic_dynamics',QA=ROOT+'/docs/papers211_215_sequence/qa/';
const OWN=QA+'p212_author_pair_binding01',PREP=QA+'p212_runtime_preparation01',DISC=QA+'p212_runtime_discovery01',PAPER=ROOT+'/papers/212-closed-pointer-orbits';
const ADOPT=QA+'p212_canonical_adoption_root01',SOURCE=QA+'p212_author_pair_source_reception01',RECEIVER=QA+'p212_pair_receiver_source_root01';
const ATTEMPT=QA+'root_replays/p212_author_pair_01',CACHE=OWN+'/never_created_root_capture_cache';
const ENV={PATH:'/usr/bin:/bin',LANG:'C.UTF-8',LC_ALL:'C.UTF-8',TZ:'UTC'};const inputs={};let checks=0;
function need(v,l){checks++;a.ok(v,l);}function same(x,y,l){checks++;a.deepStrictEqual(x,y,l);}
const pin=b=>({bytes:b.length,sha256:crypto.createHash('sha256').update(b).digest('hex')});
function read(p,k){need(path.normalize(p)===p&&path.isAbsolute(p),'literal bounded input');const l=fs.lstatSync(p),s=fs.statSync(p);need(s.isFile(),'regular resolved original');const raw=fs.readFileSync(p),v={...pin(raw),resolved:fs.realpathSync(p),symlink:l.isSymbolicLink()?fs.readlinkSync(p):null};if(inputs[p])same(v,inputs[p],'complete repeated key');if(k)for(const f of ['bytes','sha256','resolved','symlink'])if(Object.hasOwn(k,f))same(v[f],k[f],'complete supplied '+f);inputs[p]=v;return raw;}
const obj=p=>JSON.parse(read(p));
function absent(p){try{fs.lstatSync(p);throw Error('unexpected output '+p);}catch(e){if(e.code!=='ENOENT')throw e;}return {path:p,lexists:false,observed_utc:new Date().toISOString()};}
function write(n,b){fs.writeFileSync(OWN+'/'+n,b,{flag:'wx',mode:0o600});}function json(n,v){write(n,JSON.stringify(v,null,2)+'\n');}
same(Object.fromEntries(Object.entries(process.env)),ENV,'actual literal ENV4');same(process.cwd(),ROOT,'actual workspace cwd');
const beforeAbsence=[ATTEMPT,OWN+'/entry01',OWN+'/BINDING.json',PAPER+'/canonical.stdout.json',CACHE].map(absent);
same(fs.realpathSync(path.dirname(ATTEMPT)),path.dirname(ATTEMPT),'physical existing attempt parent');
function sealed(base,count){const raw=read(base+'/SHA256SUMS');need(raw.at(-1)===10,'complete nonself seal');const expected=[];for(const line of raw.toString().trimEnd().split('\n')){const m=line.match(/^([0-9a-f]{64})  (.+)$/);need(m&&!path.isAbsolute(m[2])&&!m[2].split('/').includes('..')&&m[2]!=='SHA256SUMS'&&!expected.includes(m[2]),'literal nonself row');expected.push(m[2]);read(base+'/'+m[2],{sha256:m[1]});}const actual=[],dirs=[];function walk(d){for(const n of fs.readdirSync(d)){const p=d+'/'+n,s=fs.lstatSync(p);need(!s.isSymbolicLink(),'ordinary accepted tree');if(s.isDirectory()){dirs.push(path.relative(base,p));walk(p);}else{need(s.isFile(),'ordinary accepted leaf');actual.push(path.relative(base,p));}}}walk(base);same(expected.length,count,'complete accepted payload count');same(actual.sort(),[...expected,'SHA256SUMS'].sort(),'complete accepted tree');const wanted=new Set();for(const n of expected){let d=path.dirname(n);while(d!=='.'){wanted.add(d);d=path.dirname(d);}}same(dirs.sort(),[...wanted].sort(),'complete directory membership');return {payloads:count,seal:pin(raw)};}
const acceptedPackages={adoption:sealed(ADOPT,78),controller_source:sealed(SOURCE,7),receiver_source:sealed(RECEIVER,7)};
same(obj(SOURCE+'/RESULT.json').status,'PASS_ROOT_P212_STRICT_PAIR_CONTROLLER_SOURCE_RECEPTION_ONLY');
same(obj(RECEIVER+'/RESULT.json').status,'PASS_ROOT_COMPLETE_P212_PAIR_RECEIVER_SOURCE_AND_STARTUP_AMENDMENT');
for(const p of [SOURCE+'/INPUTS.json',RECEIVER+'/INPUTS.json',ADOPT+'/ADOPTION_CLOSING_INPUTS.json'])for(const[q,k]of Object.entries(obj(p)))read(q,k);
const template=obj(PREP+'/PAIR.pending.json'),b=structuredClone(template);same(b.approved,false);same(b.mode,'pair');
const parameters=obj(PAPER+'/PARAMETERS.json');same(parameters,b.schema.equalities.find(x=>JSON.stringify(x.path)==='["parameters"]').value,'whole typed parameters');
for(const k of b.capsule_files)read(k.path,k);for(const[p,k]of Object.entries(b.adapter_sources))read(p,k);
const scientific=read(PAPER+'/verify.py').toString();same(scientific.split('\n').filter(l=>/^(import |from )/.test(l)),['import itertools','import json','import math','import sys','from fractions import Fraction'],'unchanged whole literal import surface');
const canonical=read(PAPER+'/CANONICAL.json',{bytes:12501943,sha256:'1a32be6bb36dd8b4b27b9891a42876fe235bf2555a3dd9f4e36195a6ad72676c'});
const adoption=obj(ADOPT+'/adoption01/RESULT.json');need(canonical.equals(read(adoption.source)),'whole accepted original output/canonical');
const controller=read(OWN+'/run.py',{bytes:6263,sha256:'2c6f2a7378a7c5411b262376415e1a77bbdc35523eb7821e44b01a787da93fd0'});
need(controller.equals(read(QA+'p212_author_pair_controller_preparation01/run.py')),'whole physical controller source');
const deployment=obj(OWN+'/DEPLOYMENT_NATIVE.json');same(deployment.length,6);for(const r of deployment){same(r.result.exit_code,0);same(r.result.output,'');}
const guardPath=ADOPT+'/guard_runtime.js';read(guardPath,{bytes:3913,sha256:'6819afc1081b5b5cbd64cca958f0827b29e66205c6a17f8de1548b48c6b87933'});
const guard=require(guardPath),lock=obj(DISC+'/RUNTIME_LOCK.json');
const runtime=guard(lock,p=>{const raw=read(p);return {raw,key:inputs[p]};});same(runtime.checks,1241,'complete accepted current runtime/config guard');json('RUNTIME_AT_BINDING.json',runtime);json('ROOT_ABSENCE_AT_BINDING.json',{scope:'Actual preparation-time observations, not yet immediate pre-production or continuous tracing',observations:beforeAbsence});
b.approved=true;b.attempt=ATTEMPT;b.reviewed_static_source_import_closure=true;b.reviewed_schema_and_parameters=true;b.reviewed_compile_exec_interface=true;b.timeouts={science:300,native:60,envelope:900};
b.runtime_lock={path:DISC+'/RUNTIME_LOCK.json',...inputs[DISC+'/RUNTIME_LOCK.json']};b.canonical={path:PAPER+'/CANONICAL.json',...inputs[PAPER+'/CANONICAL.json']};
const selected=[OWN+'/AUTHORITY.md',OWN+'/SOURCE_CALL_INTERFACE.md',OWN+'/ROOT_STARTUP_CONTRACT.md',OWN+'/run.py',OWN+'/prepare_binding.js',OWN+'/DEPLOYMENT_NATIVE.json',OWN+'/RUNTIME_AT_BINDING.json',OWN+'/ROOT_ABSENCE_AT_BINDING.json',
 QA+'p212_author_initial_binding01/SOURCE_CALL_INTERFACE.md',QA+'p212_author_initial_binding01/prepare_binding.js',QA+'p212_author_source_reception/RECEPTION.md',QA+'p212_author_source_reception/ROLE_DECISION.md',QA+'p212_runtime_source_root_reception01/RECEPTION.md',QA+'p212_runtime_discovery_root_reception01/RECEPTION.md',QA+'p212_runtime_discovery_root_reception01/SHA256SUMS',
 ADOPT+'/RECEPTION.md',ADOPT+'/ADOPTION_CLOSING_RESULT.json',ADOPT+'/ADOPTION_CLOSING_INPUTS.json',ADOPT+'/SHA256SUMS',guardPath,SOURCE+'/RECEPTION.md',SOURCE+'/RESULT.json',SOURCE+'/INPUTS.json',SOURCE+'/SHA256SUMS',RECEIVER+'/RECEPTION.md',RECEIVER+'/RESULT.json',RECEIVER+'/INPUTS.json',RECEIVER+'/SHA256SUMS',
 QA+'p212_saved_output_root_reception01/RECEPTION.md',QA+'p212_saved_output_root_reception01/EXECUTION_RECEPTION_RESULT.json',QA+'p212_saved_output_root_reception01/EXECUTION_RECEPTION_INPUTS.json',QA+'p212_saved_output_root_reception01/SHA256SUMS',PREP+'/PLAN.md',PREP+'/BINDING_CONTRACT.md',PREP+'/SHA256SUMS',PAPER+'/OUTPUT_SCHEMA.md',PAPER+'/OUTPUT_PLAN.md'];
same(new Set(selected).size,selected.length,'no duplicate provenance roles');for(const p of selected)read(p);b.provenance_inputs=selected.map(p=>({path:p,...inputs[p]}));
b.pending_obligations=['Actual separately bound strict pair and three native raw comparisons','Complete actual native/dependency/configuration/startup reception and unchanged-key canonical semantic reuse','Initial build, manuscript reviews and freezes, terminal builds/views remain separate'];
b.root_authority_decision='AUTHORIZE_ONE_P212_AUTHOR_STRICT_PAIR_NO_ADOPTION';b.root_execution_interface='pinned compile/exec in a fresh isolated child, exact main/file/argv/capsule cwd; not direct python verify.py';
b.root_capture={entry:OWN+'/entry01',timeout_seconds:1200,environment:ENV,controller:{path:OWN+'/run.py',...inputs[OWN+'/run.py']},outer_argv_without_binding_digest:['/usr/bin/python3.10','-I','-S','-B','-X','pycache_prefix='+ATTEMPT+'/never_created_outer_cache',PREP+'/p212_runtime.py','outer',OWN+'/BINDING.json',ATTEMPT]};
const reverse=structuredClone(b);for(const k of ['approved','attempt','reviewed_static_source_import_closure','reviewed_schema_and_parameters','reviewed_compile_exec_interface','timeouts','canonical','runtime_lock','provenance_inputs','pending_obligations'])reverse[k]=structuredClone(template[k]);for(const k of ['root_authority_decision','root_execution_interface','root_capture'])delete reverse[k];same(reverse,template,'whole exact reverse delta');
for(const p of Object.keys(inputs))read(p,inputs[p]);same(guard(lock,p=>{const raw=read(p);return {raw,key:inputs[p]};}),runtime,'whole runtime/config state unchanged during binding');
for(const p of [ATTEMPT,OWN+'/entry01',PAPER+'/canonical.stdout.json',CACHE])absent(p);
json('INPUTS_AT_BINDING.json',inputs);json('BINDING.json',b);
const result={status:'ROOT_P212_PAIR_BINDING_READY_NOT_EXECUTED',binding:{path:OWN+'/BINDING.json',...pin(fs.readFileSync(OWN+'/BINDING.json'))},checks,input_paths:Object.keys(inputs).length,runtime_keys:130,resolved_runtime_keys:121,configuration_paths:69,membership_paths:292,loader_paths:9,full_guard_checks:1241,provenance_count:selected.length,accepted_packages:acceptedPackages,attempt:ATTEMPT,entry:OWN+'/entry01',scientific_executions:0,canonical_absent:false,canonical_unchanged:true,root_startup_prefix:CACHE,root_prefix_absent_observed:true,reverse_delta_passed:true};
json('RESULT.json',result);console.log(JSON.stringify(result));
